File: Backend/model/src/main/java/net/chrisparton/sparkled/model/render/Led.java
package net.chrisparton.sparkled.model.render;

import java.awt.*;
import java.util.Objects;

public class Led {

    private static final int R = 0;
    private static final int G = 1;
    private static final int B = 2;

    private final byte[] ledData;
    private final int ledNumber;
    private final int index;

    public Led(byte[] ledData, int ledNumber) {
        this.ledData = ledData;
        this.ledNumber = ledNumber;
        this.index = ledNumber * 3;
    }

    public void addColor(Color color) {
        addRgb(color.getRed(), color.getGreen(), color.getBlue());
    }

    public void addRgb(int r, int g, int b) {
        setR(Math.min(getR() + r, 255));
        setG(Math.min(getG() + g, 255));
        setB(Math.min(getB() + b, 255));
    }

    public int getLedNumber() {
        return ledNumber;
    }

    public int getR() {
        return ledData[index + R] & 0xFF;
    }

    private void setR(int r) {
        ledData[index + R] = (byte) r;
    }

    public int getG() {
        return ledData[index + G] & 0xFF;
    }

    private void setG(int g) {
        ledData[index + G] = (byte) g;
    }

    public int getB() {
        return ledData[index + B] & 0xFF;
    }

    private void setB(int b) {
        ledData[index + B] = (byte) b;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        } else if (!(o instanceof Led)) {
            return false;
        }

        Led led = (Led) o;
        return getR() == led.getR() && getG() == led.getG() && getB() == led.getB();
    }

    @Override
    public int hashCode() {
        return Objects.hash(getR(), getG(), getB());
    }

    @Override
    public String toString() {
        return String.format("#%02X%02X%02X", getR(), getG(), getB());
    }
}

File: model/src/main/kotlin/io/sparkled/model/render/Led.kt
package io.sparkled.model.render

import io.sparkled.model.util.MathUtils.lerp
import java.awt.Color
import kotlin.math.roundToInt

// TODO consider replacing this class with functions to save memory
data class Led(
    @Suppress("ArrayInDataClass") private val ledData: ByteArray,
    private val pixelIndex: Int,
    private val offset: Int,
) {
    private val index: Int = pixelIndex * BYTES_PER_LED

    fun setColor(color: Color, alpha: Float) {
        this.r = lerp(this.r.toFloat(), color.red.toFloat(), alpha).roundToInt()
        this.g = lerp(this.g.toFloat(), color.green.toFloat(), alpha).roundToInt()
        this.b = lerp(this.b.toFloat(), color.blue.toFloat(), alpha).roundToInt()
    }

    fun addColor(color: Color) = addRgb(color.red, color.green, color.blue)
    fun subtractColor(color: Color) = addRgb(-color.red, -color.green, -color.blue)
    fun maskColor(color: Color) = addRgb(-(255 - color.red), -(255 - color.green), -(255 - color.blue))

    fun setRgb(r: Int, g: Int, b: Int) {
        this.r = r.coerceIn(0, 255)
        this.g = g.coerceIn(0, 255)
        this.b = b.coerceIn(0, 255)
    }

    fun addRgb(r: Int, g: Int, b: Int) {
        this.r = (this.r + r).coerceIn(0, 255)
        this.g = (this.g + g).coerceIn(0, 255)
        this.b = (this.b + b).coerceIn(0, 255)
    }

    private var r: Int
        get() = (ledData[offset + index + R]).toInt() and 0xFF
        private set(r) {
            ledData[offset + index + R] = r.toByte()
        }

    private var g: Int
        get() = (ledData[offset + index + G]).toInt() and 0xFF
        private set(g) {
            ledData[offset + index + G] = g.toByte()
        }

    private var b: Int
        get() = (ledData[offset + index + B]).toInt() and 0xFF
        private set(b) {
            ledData[offset + index + B] = b.toByte()
        }

    override fun toString(): String {
        return String.format("#%02X%02X%02X", r, g, b)
    }

    companion object {
        const val BYTES_PER_LED = 3

        private const val R = 0
        private const val G = 1
        private const val B = 2
    }
}
